app/scripts/lib/stores/persistence-manager.ts

```diff
@@ -57,58 +97,158 @@ export class PersistenceManager {
    * includes a single key which is 'version' and contains the current version
    * number of the state tree.
    */
-  #metadata?: { version: number };
+  #metadata?: MetaData;
 
   #isExtensionInitialized: boolean = false;
 
-  #localStore: ExtensionStore | ReadOnlyNetworkStore;
+  #localStore: BaseStore;
 
-  constructor({
-    localStore,
-  }: {
-    localStore: ExtensionStore | ReadOnlyNetworkStore;
-  }) {
+  #backupDb: IndexedDBStore;
+
+  #backup?: string;
+
+  #open: boolean = false;
+
+  constructor({ localStore }: { localStore: BaseStore }) {
     this.#localStore = localStore;
+    this.#backupDb = new IndexedDBStore();
   }
 
-  setMetadata(metadata: { version: number }) {
+  async open() {
+    if (!this.#open) {
+      await this.#backupDb.open('metamask-backup', 1);
+      this.#open = true;
+    }
+  }
+
+  setMetadata(metadata: MetaData) {
     this.#metadata = metadata;
   }
 
+  /**
+   * Sets the state in the local store.
+   *
+   * @param state - The state to set in the local store. This should be an object
+   * containing the state data to be stored.
+   * @throws Error if the state is missing or if the metadata is not set before
+   * calling this method.
+   * @throws Error if the local store is not open.
+   * @throws Error if the data persistence fails during the write operation.
+   */
   async set(state: MetaMaskStateType) {
+    await this.open();
+
     if (!state) {
       throw new Error('MetaMask - updated state is missing');
     }
-    if (!this.#metadata) {
+    const meta = this.#metadata;
+    if (!meta) {
       throw new Error('MetaMask - metadata must be set before calling "set"');
     }
-    try {
-      await this.#localStore.set({ data: state, meta: this.#metadata });
-      if (this.#dataPersistenceFailing) {
-        this.#dataPersistenceFailing = false;
-      }
-    } catch (err) {
-      if (!this.#dataPersistenceFailing) {
-        this.#dataPersistenceFailing = true;
-        captureException(err);
-      }
-      log.error('error setting state in local store:', err);
-    } finally {
-      this.#isExtensionInitialized = true;
-    }
+
+    await navigator.locks.request(
+      STATE_LOCK,
+      { mode: 'exclusive' },
+      async () => {
+        try {
+          // atomically set all the keys
+          await this.#localStore.set({
+            data: state,
+            meta,
+          });
+
+          const backup = makeBackup(state);
+          // if we have a vault we can back it up
+          if (backup.KeyringController?.vault) {
```

Comment:
Nit: Maybe we could loosen the parameter type of `hasVault`, and use that here as well?

This is perfectly adequate as-is, but if you accept my earlier suggestion about the Backup type, this will require an additional check to ensure `vault` exists as a property. And in that case, using `hasVault` would be easiest probably.

Reply:
done